src/server/memory_cmd.cc

```diff
@@ -143,6 +153,110 @@ void MemoryCmd::Run(CmdArgList args) {
+      } else {
+        mem.replication_connection_count++;
+        mem.replication_memory += dfly_conn->GetMemoryUsage();
+      }
+
+      if (cntx != nullptr) {
+        mem.pipelined_bytes += cntx->conn_state.exec_info.body.capacity() * sizeof(StoredCmd);
+        for (const auto& pipeline : cntx->conn_state.exec_info.body) {
+          mem.pipelined_bytes += pipeline.UsedHeapMemory();
+        }
+      }
+    });
+  }
+
+  return mem;
+}
+
+void PushMemoryUsageStats(const base::IoBuf::MemoryUsage& mem, string_view prefix, size_t total,
+                          vector<string>* stats) {
+  stats->push_back(absl::StrCat(prefix, ".total_bytes"));
+  stats->push_back(absl::StrCat(total));
+  stats->push_back(absl::StrCat(prefix, ".consumed_bytes"));
+  stats->push_back(absl::StrCat(mem.consumed));
+  stats->push_back(absl::StrCat(prefix, ".pending_input_bytes"));
+  stats->push_back(absl::StrCat(mem.input_length));
+  stats->push_back(absl::StrCat(prefix, ".pending_output_bytes"));
+  stats->push_back(absl::StrCat(mem.append_length));
+}
+
+}  // namespace
+
+void MemoryCmd::Stats() {
+  vector<string> stats;
+  stats.reserve(50);
+  auto server_metrics = owner_->GetMetrics();
+
+  // RSS
+  stats.push_back("rss_bytes");
+  stats.push_back(absl::StrCat(rss_mem_current.load()));
+  stats.push_back("rss_peak_bytes");
+  stats.push_back(absl::StrCat(rss_mem_peak.load()));
+
+  // Used by DbShards and DashTable
+  stats.push_back("data_bytes");
+  stats.push_back(absl::StrCat(used_mem_current.load()));
+  stats.push_back("data_peak_bytes");
+  stats.push_back(absl::StrCat(used_mem_peak.load()));
+
+  ConnectionMemoryUsage connection_memory = GetConnectionMemoryUsage(owner_);
+
+  // Connection stats, excluding replication connections
+  stats.push_back("connections.count");
+  stats.push_back(absl::StrCat(connection_memory.connection_count));
+  PushMemoryUsageStats(
+      connection_memory.connections_memory, "connections",
+      connection_memory.connections_memory.GetTotalSize() + connection_memory.pipelined_bytes,
+      &stats);
+  stats.push_back("connections.pipeline_bytes");
+  stats.push_back(absl::StrCat(connection_memory.pipelined_bytes));
+
+  // Replication connection stats
+  stats.push_back("replication.connections_count");
+  stats.push_back(absl::StrCat(connection_memory.replication_connection_count));
+  PushMemoryUsageStats(connection_memory.replication_memory, "replication",
+                       connection_memory.replication_memory.GetTotalSize(), &stats);
+
+  Mutex mu;
+  size_t serialization_memory ABSL_GUARDED_BY(mu);
+  shard_set->pool()->AwaitFiberOnAll([&](auto*) {
+    lock_guard lock(mu);
+    serialization_memory += SliceSnapshot::GetThreadLocalMemoryUsage();
+  });
+
+  // Serialization stats, including both replication-related serialization and saving to RDB files.
+  stats.push_back("serialization");
+  stats.push_back(absl::StrCat(serialization_memory));
+
+  return (*cntx_)->SendSimpleStrArr(stats);
```

Comment:
nice that you used a formatted output. nit: you can specify this as MAP to get even better representation for resp3 enabled clients.